### PR TITLE
Feature: Added federated repo type

### PR DESCRIPTION
**Pull request by dakky**
## Description

Artifactory provides in a new Type of Repo: `federated`. this is basically a local repo, except that the model has a field members, which contains URLs to other artifactory instances and replicated (bi-directional) the content between these instances

Fixes # (issue)
-

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How has it been tested ?

- [x] pytest ;)
- [x] Integration test:
  ```
  from __future__ import annotations

  from pyartifactory import Artifactory
  from pyartifactory.models import FederatedRepository, LocalRepository

  art = Artifactory(
      url="https://URL/artifactory",
      auth=("XXX", "XXX"),
      verify=False,
  )

  # testing of "old" behavior still works fine
  local_repo = LocalRepository(key="local_test")
  new_local_repo = art.repositories.create_repo(local_repo)

  # testing if federated repo creation works
  fed_repo = FederatedRepository(key="test2")
  new_fed_repo = art.repositories.create_repo(fed_repo)
  ```

## Checklist:

- [x] My PR is ready for prime time! Otherwise use the ["Draft PR" feature](https://help.github.com/en/articles/about-pull-requests#draft-pull-requests)
- [x] All commits have a correct title
- [x] Readme has been updated
- [x] Quality tests are green (see Codacy)
- [x] Automated tests are green (see pipeline)

### CODE REVIEW

File: pyartifactory/objects/repository.py [hunk 1]
```diff
@@ -35,18 +37,32 @@ def get_repo(self, repo_name: str) -> AnyRepositoryResponse:
...
+                return VirtualRepositoryResponse.model_validate(response_data)
+            elif rclass == RClassEnum.remote:
+                return RemoteRepositoryResponse.model_validate(response_data)
+            elif rclass == RClassEnum.federated:
+                return FederatedRepositoryResponse.model_validate(response_data)
+            else:
+                # this should never happen and is a missing repotype in the library
+                raise ArtifactoryError(
+                    f"Unknown repository type found in response: {rclass}. Please report this issue.",
+                )
```
Comment (dakky) [inline pyartifactory/objects/repository.py:65]: I rebuilt this section as another repo type requires another nested section of try/catch and this was hard to read/maintain.  Functionality should be the same.

File: pyartifactory/models/repository.py [hunk 2]
```diff
@@ -306,3 +312,54 @@ class RemoteRepositoryResponse(RemoteRepository):
...
+    enableCocoaPodsSupport: bool = False
+    enableConanSupport: bool = False
+    enableDebianSupport: bool = False
+    enablePypiSupport: bool = False
+    enablePuppetSupport: bool = False
+    enableDockerSupport: bool = False
+    forceNugetAuthentication: bool = False
+    enableVagrantSupport: bool = False
+    enableGitLfsSupport: bool = False
+    enableDistRepoSupport: bool = False
```
Comment (dakky) [inline pyartifactory/models/repository.py:365]: A little bit ugly as this is basically a copy of the local repo extended with the member field

File: pyartifactory/objects/repository.py (hunk 1 above)
Comment (dakky) [reply pyartifactory/objects/repository.py:65]: ... except, that now there is a error thrown when rcall does not match instead of a magic "i don't know the rcall enum. I think its a remote repo"

File: pyartifactory/models/repository.py [hunk 3]
```diff
@@ -114,6 +115,11 @@ class ContentSynchronisation(BaseModel):
     source: Source = Source()
 
 
+class FederatedMembers(BaseModel):
+    url: str = ""
+    enabled: bool = False
```
Comment (anancarv) [inline pyartifactory/models/repository.py:120]: From what I can see in the [Artifactory API](https://jfrog.com/help/r/jfrog-rest-apis/repository-configuration-json), enabled is a str, not a boolean.

```
"members": [{"url": "http://targetartifactory/artifactory/repositoryName", "enabled":"true"},
```

File: pyartifactory/models/repository.py [hunk 4]
```diff
@@ -306,3 +312,54 @@ class RemoteRepositoryResponse(RemoteRepository):
...
+    yumRootDepth: int = 0
+    enableFileListsIndexing: bool = False
+    optionalIndexCompressionFormats: Optional[List[str]] = None
+    downloadRedirect: bool = False
+    cdnRedirect: bool = False
+    blockPushingSchema1: bool = False
+    primaryKeyPairRef: Optional[str] = None
+    secondaryKeyPairRef: Optional[str] = None
+    priorityResolution: bool = False
+    cargoInternalIndex: bool = False
```
Comment (anancarv) [inline pyartifactory/models/repository.py:345]: If the `FederatedRepository` is just a LocalRepository but with some additional fields as you mentioned. Why not do something like this ?

```
class FederatedRepository(LocalRepository):
       rclass: Literal[RClassEnum.federated] = RClassEnum.federated
       members: List[FederatedMembers] = []
```

Same for FederatedRepositoryResponse:
```
class FederatedRepository(LocalRepositoryResponse):
       rclass: Literal[RClassEnum.federated] = RClassEnum.federated
       members: List[FederatedMembers] = []
```

Doing this would reduce a lot of code duplication

File: pyartifactory/objects/repository.py (hunk 1 above)
Comment (anancarv) [reply pyartifactory/objects/repository.py:65]: I like that new approach. It's way more maintainable 👍🏿

File: tests/test_repositories.py [hunk 5]
```diff
@@ -42,6 +44,19 @@
     description="updated",
 )
 
+FEDERATED_REPOSITORY = RemoteRepository(key="test_federated_repository", url="http://test-url.com")
```
Comment (anancarv) [inline tests/test_repositories.py:47]: I think the error you were talking about is because of that line. You wan to create a FEDERATED_REPO but, using thec class `RemoteRepository`

File: tests/test_repositories.py [hunk 6]
```diff
@@ -42,6 +44,19 @@
     description="updated",
 )
 
+FEDERATED_REPOSITORY = RemoteRepository(key="test_federated_repository", url="http://test-url.com")
+FEDERATED_REPOSITORY_RESPONSE = FederatedRepositoryResponse(key="test_remote_repository", url="http://test-url.com")
```
Comment (anancarv) [inline tests/test_repositories.py:48]: Where is the key members ?

File: tests/test_repositories.py [hunk 7]
```diff
@@ -42,6 +44,19 @@
     description="updated",
 )
 
+FEDERATED_REPOSITORY = RemoteRepository(key="test_federated_repository", url="http://test-url.com")
+FEDERATED_REPOSITORY_RESPONSE = FederatedRepositoryResponse(key="test_remote_repository", url="http://test-url.com")
+UPDATED_FEDERATED_REPOSITORY = FederatedRepository(
+    key="test_federated_repository",
+    url="http://test-url.com",
```
Comment (anancarv) [inline tests/test_repositories.py:51]: Same here. Where is the key members ?

File: tests/test_repositories.py [hunk 8]
```diff
@@ -42,6 +44,19 @@
...
+FEDERATED_REPOSITORY = RemoteRepository(key="test_federated_repository", url="http://test-url.com")
+FEDERATED_REPOSITORY_RESPONSE = FederatedRepositoryResponse(key="test_remote_repository", url="http://test-url.com")
+UPDATED_FEDERATED_REPOSITORY = FederatedRepository(
+    key="test_federated_repository",
+    url="http://test-url.com",
+    description="updated",
+)
+UPDATED_FEDERATED_REPOSITORY_RESPONSE = FederatedRepositoryResponse(
+    key="test_federated_repository",
+    url="http://test-url.com",
```
Comment (anancarv) [inline tests/test_repositories.py:56]: Members missing

File: tests/test_repositories.py [hunk 9]
```diff
@@ -170,6 +204,29 @@ def test_create_remote_repository_success(mocker):
...
+    responses.add(
+        responses.PUT,
+        f"{URL}/api/repositories/{FEDERATED_REPOSITORY.key}",
+        json=FEDERATED_REPOSITORY_RESPONSE.model_dump(),
+        status=201,
+    )
+    responses.add(
+        responses.GET,
+        f"{URL}/api/repositories/{FEDERATED_REPOSITORY.key}",
+        json=FEDERATED_REPOSITORY_RESPONSE.model_dump(),
```
Comment (anancarv) [inline tests/test_repositories.py:219]: I don't think those tests work since `FEDERATED_REPOSITORY_RESPONSE` is missing the `members` key

**Review by anancarv — CHANGES_REQUESTED** [review]

**Comment by anancarv** [comment]
Thanks @dakky for your contribution, sorry for the delay for reviewing your changes. There are just a few things to change for the MR to be approved

File: pyartifactory/models/repository.py [hunk 10]
```diff
@@ -114,6 +115,11 @@ class ContentSynchronisation(BaseModel):
     source: Source = Source()
 
 
+class FederatedMembers(BaseModel):
+    url: str = ""
+    enabled: bool = True
```
Comment (anancarv) [inline pyartifactory/models/repository.py:120]: Are you sure about that boolean type here ? In the official API , enabled is a string not a boolean (See [documentation](https://jfrog.com/help/r/jfrog-rest-apis/repository-configuration-json))

File: pyartifactory/models/repository.py [hunk 11]
```diff
@@ -306,3 +312,16 @@ class RemoteRepositoryResponse(RemoteRepository):
...
+    """Models a federated Repository."""
+
+    rclass: Literal[RClassEnum.federated] = RClassEnum.federated
+    members: List[FederatedMembers] = []
+
+
+class FederatedRepositoryResponse(FederatedRepository):
+    """Models a federated repository response."""
+
+    members: List[FederatedMembers] = []
```
Comment (anancarv) [inline pyartifactory/models/repository.py:327]: No need of that if you're inheriting from `FederatedRepository`. I wonder if there are other hidden params (not documented on the API) returned. Can you please share the result of a FederatedRepositoryResponse using the API ?

File: pyartifactory/objects/repository.py [hunk 12]
```diff
@@ -35,42 +33,48 @@ def get_repo(self, repo_name: str) -> AnyRepositoryResponse:
...
-    def create_repo(self, repo: LocalRepository) -> LocalRepositoryResponse:
-        ...
-
-    @overload
-    def create_repo(self, repo: VirtualRepository) -> VirtualRepositoryResponse:
-        ...
-
-    @overload
-    def create_repo(self, repo: RemoteRepository) -> RemoteRepositoryResponse:
-        ...
```
Comment (anancarv) [inline pyartifactory/objects/repository.py:67]: Why did you remove that ? It is important for the typing. Please revert it and add a section for federated repositories

File: pyartifactory/objects/repository.py [hunk 13]
```diff
@@ -87,19 +91,10 @@ def create_repo(self, repo: AnyRepository) -> AnyRepositoryResponse:
...
-    def update_repo(self, repo: LocalRepository) -> LocalRepositoryResponse:
-        ...
-
-    @overload
-    def update_repo(self, repo: VirtualRepository) -> VirtualRepositoryResponse:
-        ...
-
-    @overload
-    def update_repo(self, repo: RemoteRepository) -> RemoteRepositoryResponse:
-        ...
```
Comment (anancarv) [inline pyartifactory/objects/repository.py:100]: Same here

**Review by anancarv — CHANGES_REQUESTED** [review]

File: pyartifactory/models/repository.py (hunk 3 above)
Comment (dakky) [reply pyartifactory/models/repository.py]: this is true. The "problem" is: the repsonse object returns this field as boolean.
So basically there are two ways:
1. keep it as it is. It'S not 100% clean but works as the api seems to accept a boolean too
2.  model an additional reposonse model:
   ```
   class FederatedMembers(BaseModel):
    url: str = ""
    # enabled: bool = True
    enabled: str = "true"

class FederatedMembersResponse(BaseModel):
    url: str = ""
    enabled: bool = True
    ```
    This seems a little bit weird as it's only nessecary because of an unclean design of the api

File: tests/test_repositories.py (hunk 5 above)
Comment (dakky) [reply tests/test_repositories.py]: this is correct. it's wrong. although it does not cause the mypy errors ;)

File: tests/test_repositories.py (hunk 7 above)
Comment (dakky) [reply tests/test_repositories.py:59]: same as above

File: tests/test_repositories.py (hunk 8 above)
Comment (dakky) [reply tests/test_repositories.py:65]: see above :-D

File: tests/test_repositories.py (hunk 9 above)
Comment (dakky) [reply tests/test_repositories.py:229]: aand again :D see above

File: pyartifactory/objects/repository.py (hunk 12 above)
Comment (dakky) [reply pyartifactory/objects/repository.py:67]: are you shure that it is nessecary?
1. it simply breaks mypy because local and feaderated signatures are too similar as mypy could handle this?
2. i THINK (with my limited knowledge of python typing) these where simply not nessecary because the typing is handled by the `AnyRepository` in models/__init__.py?

Comment (dakky) [reply pyartifactory/objects/repository.py:67]: the error when simply adding a section for the federated repo is:
```
pyartifactory/objects/repository.py:74: error: Overloaded function signatures 1
and 2 overlap with incompatible return types  [misc]
        def create_repo(self, repo: FederatedRepository) -> FederatedRepos...
        ^
pyartifactory/objects/repository.py:115: error: Overloaded function signatures
1 and 2 overlap with incompatible return types  [misc]
        def update_repo(self, repo: FederatedRepository) -> FederatedRepos...
        ^
```

File: pyartifactory/models/repository.py (hunk 11 above)
Comment (dakky) [reply pyartifactory/models/repository.py:327]:
```
FederatedRepository(key='test2'
 projectKey=None
 rclass=<RClassEnum.federated: 'federated'>
 packageType=<PackageTypeEnum.generic: 'generic'>
 description=None
 notes=None
 includesPattern='**/*'
 excludesPattern=''
 repoLayoutRef='maven-2-default'
 checksumPolicyType=<ChecksumPolicyType.client_checksums: 'client-checksums'>
 handleReleases=True
 handleSnapshots=True
 maxUniqueSnapshots=0
 maxUniqueTags=0
 debianTrivialLayout=False
 snapshotVersionBehavior=<SnapshotVersionBehavior.non_unique: 'non-unique'>
 suppressPomConsistencyChecks=False
 blackedOut=False
 xrayIndex=False
 propertySets=None
 dockerApiVersion='V2'
 archiveBrowsingEnabled=False
 calculateYumMetadata=False
 yumRootDepth=0
 enableFileListsIndexing=False
 optionalIndexCompressionFormats=None
 downloadRedirect=False
 cdnRedirect=False
 blockPushingSchema1=False
 primaryKeyPairRef=None
 secondaryKeyPairRef=None
 priorityResolution=False
 cargoInternalIndex=False
 members=[FederatedMembers(url='https://testdomain.de/artifactory/test2'
 enabled=True)])
```

File: pyartifactory/objects/repository.py (hunk 12 above)
Comment (dakky) [reply pyartifactory/objects/repository.py:67]: when looking at the code, I'm not shure if "reducing the code dupication in the model" is worth the hassle in the code to satisfy the typing

File: pyartifactory/models/repository.py (hunk 3 above)
Comment (anancarv) [reply pyartifactory/models/repository.py]: I would go for option 2. And to ensure people write either "true" or "false", I would add a literal:
```
enabled: Literal["true", "false"] = "true"
```

File: pyartifactory/objects/repository.py (hunk 12 above)
Comment (anancarv) [reply pyartifactory/objects/repository.py:67]: > i THINK (with my limited knowledge of python typing) these where simply not nessecary because the typing is handled by the AnyRepository in models/__init__.py?

With the following code:
```
    def create_repo(
        self,
        repo: AnyRepository,
    ) -> AnyRepositoryResponse:
```

We take input a `AnyRepository` and return a `AnyRepositoryResponse`. Which in theory means that we can have as input a `LocalRepository` and return a `VirtualRepositoryResponse `. Which is not what we want.

The overload function makes sure that `LocalRepository` returns a `LocalRepositoryResponse`.

File: pyartifactory/models/repository.py (hunk 2 above)
Comment (anancarv) [reply pyartifactory/models/repository.py]: this was actually the good way

File: pyartifactory/objects/repository.py (hunk 12 above)
Comment (dakky) [reply pyartifactory/objects/repository.py:67]: okay i'll revert this and add the duplication again. but this keeps it much more readable imho. I'll squash the intermediate commits in order to keep the history clean

**Review by anancarv — APPROVED** [review]
lgtm !

**Comment by dakky** [comment]
hmm somethings wrong with the codacy token?

**Comment by anancarv** [comment]
> hmm somethings wrong with the codacy token?

It's ok. It's a known problem. It is because you don't have the codacy token on the settings of you forked repo.
